README.md
# Whitelist AWS IP Ranges - Lambda

Lambda for whitelisting Amazon IP ranges in Security Group outbound rules.

Works by updating Security Group Egress rules with a list of AWS IP Ranges based on AWS Service Nam.
Pulls latest JSON from - https://docs.aws.amazon.com/general/latest/gr/aws-ip-ranges.html

This is written in Go as practice for using the AWS SDK and Golang in general.

## Build Lambda zip

Locally :

Linux : 
```
# Get dependency
go get -u github.com/aws/aws-lambda-go/cmd/build-lambda-zip
```

```
# Compile and zip
GOOS=linux go build main.go && zip go_lambda.zip main
```

Windows : 
```
# Get dependency
go get -u github.com/aws/aws-lambda-go/cmd/build-lambda-zip
```

```
# Compile and zip
$env:GOOS = "linux"
$env:CGO_ENABLED = "0"
$env:GOARCH = "amd64"
go build -o main main.go; ~\Go\Bin\build-lambda-zip.exe -output go_lambda.zip main
```

Using Make : 

```
$ make run
go run main.go
Hello

$ make build                                      
go build -o bin/lambda lambda.go

$ make compile
echo "Compiling for every OS and Platform"
Compiling for every OS and Platform
# Linux
GOOS=linux GOARCH=amd64 go build -o bin/lambda-linux-amd64 lambda.go
# Windows binary
GOOS=windows GOARCH=amd64 go build -o bin/lambda-windows-amd64 lambda.go
```

### Create Lambda
```
aws lambda create-function --function-name my-function --runtime go1.x \
  --zip-file fileb://go_lambda.zip --handler main \
  --role arn:aws:iam::123456789012:role/execution_role

```
Setup Lambda env variables :

|  Key | Value  | Description |
|---   |---     |---          |
| amazonIPRangesURL |	https://ip-ranges.amazonaws.com/ip-ranges.json | URL with AWS IP ranges |
| awsRegion  | eu-central-1 | AWS Region |
| dynamoTableName  | whitelistedIPRanges | Name of DynamoDB table to store whitelisted IP ranges |
| previousDateParamStore  | lastModifiedDateIPRanges | SSM Param store name to keep modified date of AWS JSON file |
| securityGroupIDs  | sg-041c5e7daf95e16a3 | Comma separated list of Security groups (no spaces) |
| servicesToBeWhitelist  | S3 | Comma separated list of AWS Services from JSON list (no spaces) |


Setup a Lambda Trigger, e.g. time based
EventBridge trigger - scheduled expressions
```
# Run every hour
cron(0 * * * ? *)

# Run every 30 min
cron(0/30 * * * ? *)
```

### IAM Policy Permissions needed
TODO: add IAM policy example
```
SSM Param Store - Get/Put Parameter
Cloudwatch - log groups
DynamoDB - create/describe table, get/put item
Security Groups - update
```


### Test locally :

```
# Go Dependencies
go get -u github.com/aws/aws-sdk-go/...
go get -u github.com/aws/aws-lambda-go/lambda
```


Setup AWS Credentials
```
# Linux
export AWS_ACCESS_KEY_ID=YOUR_AKID
export AWS_SECRET_ACCESS_KEY=YOUR_SECRET_KEY
```

```
# Windows
$env:AWS_ACCESS_KEY_ID='YOUR_AKID'
$env:AWS_SECRET_ACCESS_KEY='YOUR_SECRET_KEY'
```

Hardcoded Variables example
```
// List of Security groups to be updated
securityGroupIDs := []string{"sg-041c5e7daf95e16a3"}

// List of services to be whitelisted - e.g. AMAZON, COUDFRONT, S3, EC2, API_GATEWAY, DYNAMODB, ROUTE53_HEALTHCHECKS, CODEBUILD
servicesToBeWhitelist := []string{"S3"}

// AWS JSON URL and local download path
amazonIPRangesURL := "https://ip-ranges.amazonaws.com/ip-ranges.json"

// AWS SSM Param Store that hold the last modified date of the JSON file - format "2020-09-18-21-51-15"
previousDateParamStore := "lastModifiedDateIPRanges"

// AWS DynamoDB table to be created that will maintain a list of all whitelisted IP Ranges
dynamoTableName := "whitelistedIPRanges"

// Set AWS Region
awsRegion := "eu-central-1"
```

### Common errors

##### SSM Param Store
```
# You need a IAM policy with permissions to SSM : 
panic: AccessDeniedException: User: arn:aws:iam::111111111111:user/test is not authorized to perform: ssm:GetParameter on resource: arn:aws:ssm:eu-central-1:111111111111:parameter/lastModifiedDateIPRanges
        status code: 400, request id: 4a0ab454-176d-4bc6-9418-78529da0f944

# Make sure that no permissions boundary is also not limiting you in case you already have an IAM policy with sufficient access.
```

```
"errorMessage": "updateSecurityGroups: Cannot update security group: awsUpdateSg: UnauthorizedOperation: You are not authorized to perform this operation. Encoded authorization failure message: ... n\tstatus code: 403, request id: a2cfb4ab-b102-4eaa-bda6-cc9d326c4fa7",
"errorType": "wrapError"

```

##### Security Groups
```
# You did not provide endough Security Groups to fit all IP ranges :
panic: [ERROR]: You will need [5] Security Groups, you provided [2]
```

##### Successful run example outputs
1. When the AWS JSON file has not changed
```
START RequestId: fb9dddc4-3a0a-4c1a-bc75-a82f0f28373d Version: $LATEST
IP Ranges that need to be in whitelist: [10]
List of IP Ranges : [3.5.0.0/16 52.219.0.0/16 52.95.0.0/16 108.175.0.0/16 52.92.0.0/16 54.231.0.0/16 52.218.0.0/16 52.216.0.0/16 54.222.0.0/16 52.82.0.0/16]
Last modifed date : 2020-10-10-04-51-17
Amazon JSON file has not changed since last run, exiting ...
Lambda exeuction completed successfully
END RequestId: fb9dddc4-3a0a-4c1a-bc75-a82f0f28373d
REPORT RequestId: fb9dddc4-3a0a-4c1a-bc75-a82f0f28373d	Duration: 358.33 ms	Billed Duration: 400 ms	Memory Size: 512 MB	Max Memory Used: 55 MB	Init Duration: 124.49 ms	
```

2. When AWS JSON file has changed but there are no new IP Ranges to be whitelisted : 
```
START RequestId: 81d73f22-bb79-47ac-b151-9b7787a20b36 Version: $LATEST
IP Ranges that need to be in whitelist: [10]
List of IP Ranges : [3.5.0.0/16 52.219.0.0/16 52.95.0.0/16 108.175.0.0/16 52.92.0.0/16 54.231.0.0/16 52.218.0.0/16 52.216.0.0/16 54.222.0.0/16 52.82.0.0/16]
AWS JSON file has changed since last run, previous date: 2020-10-10-04-51-17213
Updating creation date to 2020-10-10-04-51-17
Parameter store [lastModifiedDateIPRanges] type [String] updated successfully with value [2020-10-10-04-51-17]
Modified date changed
Checking if DynamoDB table exists ...
[whitelistedIPRanges] table already exists
Checking if any IP Ranges need to be whitelisted ...
Checking item: 3.5.0.0/16
Checking item: 52.219.0.0/16
Checking item: 52.95.0.0/16
Checking item: 108.175.0.0/16
Checking item: 52.92.0.0/16
Checking item: 54.231.0.0/16
Checking item: 52.218.0.0/16
Checking item: 52.216.0.0/16
Checking item: 54.222.0.0/16
Checking item: 52.82.0.0/16
No new IP Ranges found, exiting
Lambda exeuction completed successfully
END RequestId: 81d73f22-bb79-47ac-b151-9b7787a20b36
REPORT RequestId: 81d73f22-bb79-47ac-b151-9b7787a20b36	Duration: 185.74 ms	Billed Duration: 200 ms	Memory Size: 512 MB	Max Memory Used: 56 MB	
```

3. When there are new a few new IP Addresses to be added : 
```
START RequestId: f306eedc-8341-429c-be10-6e5c6a8f5dfe Version: $LATEST
IP Ranges that need to be in whitelist: [10]
List of IP Ranges : [3.5.0.0/16 52.219.0.0/16 52.95.0.0/16 108.175.0.0/16 52.92.0.0/16 54.231.0.0/16 52.218.0.0/16 52.216.0.0/16 54.222.0.0/16 52.82.0.0/16]
AWS JSON file has changed since last run, previous date: 2020-10-10-04-51-17312
Updating creation date to 2020-10-10-04-51-17
Parameter store [lastModifiedDateIPRanges] type [String] updated successfully with value [2020-10-10-04-51-17]
Modified date changed
Checking if DynamoDB table exists ...
[whitelistedIPRanges] table already exists
Checking if any IP Ranges need to be whitelisted ...
Checking item: 3.5.0.0/16
Checking item: 52.219.0.0/16
Adding IP in Dynamo table : 52.219.0.0/16
Successfully updated dynamo table whitelistedIPRanges
Checking item: 52.95.0.0/16
Checking item: 108.175.0.0/16
Checking item: 52.92.0.0/16
Checking item: 54.231.0.0/16
Checking item: 52.218.0.0/16
Adding IP in Dynamo table : 52.218.0.0/16
Successfully updated dynamo table whitelistedIPRanges
Checking item: 52.216.0.0/16
Checking item: 54.222.0.0/16
Checking item: 52.82.0.0/16

Adding IP range [52.219.0.0/16] to Security Group [sg-041c5e7daf95e16a3]... 
IP Range added successfully

Adding IP range [52.218.0.0/16] to Security Group [sg-041c5e7daf95e16a3]... 
IP Range added successfully
Security group update finished
Successfully populated IP addresses in security groups [sg-041c5e7daf95e16a3]

Describe Security Group Egress Rules : 

 {
  Description: "not used",
  GroupId: "sg-041c5e7daf95e16a3",
  GroupName: "test1234",
  IpPermissionsEgress: [{
      FromPort: 443,
      IpProtocol: "tcp",
      IpRanges: [
        {
          CidrIp: "52.95.0.0/16"
        },
        {
          CidrIp: "108.175.0.0/16"
        },
        {
          CidrIp: "52.92.0.0/16"
        },
        {
          CidrIp: "54.231.0.0/16"
        },
        {
          CidrIp: "52.216.0.0/16"
        },
        {
          CidrIp: "54.222.0.0/16"
        },
        {
          CidrIp: "52.82.0.0/16"
        },
        {
          CidrIp: "3.5.0.0/16"
        },
        {
          CidrIp: "52.219.0.0/16"
        },
        {
          CidrIp: "52.218.0.0/16"
        }
      ],
      ToPort: 443
    }],
  OwnerId: "722377226063",
  VpcId: "vpc-342c735c"
}

Changelog.md
### backlog:

**Initial release - v1.0** :
- [x] Download Amazon IP range file and parse JSON data structure
- [x] Update list of IP ranges in Security Groups / Describe Security Groups
- [x] Work around SG limit of 60 inbound/outbound rules
- [x] Persistent way of storing JSON modified date - SSM Param Store
- [x] Better error handling
- [x] Make AWS region configurable
- [x] Update only entries that don't exist already - DynamoDB persistence
  - [x] Check if DynamoDB table exists; 
  - [x] Create DynamoDB table if doesn't exist
  - [x] Add list of IP ranges in DynamoDB table
  - [x] Only update if an entry is missing
  - [x] Create list of IPs to be added in SG from DynamoDB Table
- [x] Implement Lambda handler
- [x] Create initial release v1

**Improvements - v1.1** : 
- [x] Combine download and json parse funcs into one using decoder (no need to download the file locally)
- [x] Add Lambda trigger example in Readme
- [x] Move all vars to be taken from Lambda ENV vars instead of hardcoded
- [x] Handle dependencies as Go modules
- [x] Add CI with Github actions
- [ ] Split functions into separate packages
  - [ ] SSM Param store funcs
  - [ ] Security Group functions
  - [ ] DynamoDB functions
- [ ] Add Unit Tests
- [ ] Create SSM param store if it doesnt exist
- [ ] Move all AWS svc client duplications to an init() function - https://tutorialedge.net/golang/the-go-init-function/; we can have more than 1 init() to initialize the different svc clients
- [ ] Figure out a good way to link all SGs at the end into a single one - some sort of inheritance ?
- [ ] Add IAM policy example with minimal access needed in Readme

**Fix Bugs** :
- [x] Security group updates when IPs are less than 50 (they get duplicated in all SGs)
- [x] Dynamo update items when the table needs to be created - it adds IPs in table but but goes to case where No new IP Ranges were found and doesnt update the SGs, although the Dynamo table is completely empty. putDynamoItem() also doesn't print its success messages, although items are successfully created there. But on next Lambda run eveyrthing is fine because Dynamo table with items already exist. Seems to be related to the regex check, but not sure why it happens only when table doesn't initially excist. We are hitting this condition : case dynamodb.ErrCodeResourceNotFoundException: return true, nil


main.go
package main

import (
	"encoding/json"
	"errors"
	"fmt"
	"net/http"
	"os"
	"regexp"
	"strings"
	"time"

	"github.com/aws/aws-lambda-go/lambda"
	"github.com/aws/aws-sdk-go/aws"
	"github.com/aws/aws-sdk-go/aws/awserr"
	"github.com/aws/aws-sdk-go/aws/session"
	"github.com/aws/aws-sdk-go/service/dynamodb"
	"github.com/aws/aws-sdk-go/service/ec2"
	"github.com/aws/aws-sdk-go/service/ssm"
)

// Services - Array of AWS Services and their IP ranges
type Services struct {
	SyncToken    string    `json:"syncToken"`
	CreationDate string    `json:"createDate"`
	Prefixes     []Service `json:"prefixes"`
}

// Service - AWS Service and its IP Range used in by Amazon
type Service struct {
	IPPrefix           string `json:"ip_prefix"`
	Region             string `json:"region"`
	ServiceName        string `json:"service"`
	NetworkBorderGroup string `json:"network_border_group"`
}

func main() {
	lambda.Start(LambdaHandler)
}

// LambdaHandler - Main AWS Lambda Entrypoint
func LambdaHandler() (msg string, err error) {
	// List of Security groups to be updated
	// securityGroupIDs := []string{"sg-041c5e7daf95e16a3", "sg-00ffabccebd5efda2"}
	// Comma-serpated env var - no spaces
	securityGroupIDs := strings.Split(os.Getenv("securityGroupIDs"), ",")

	// List of services to be whitelisted - e.g. AMAZON, COUDFRONT, S3, EC2, API_GATEWAY, DYNAMODB, ROUTE53_HEALTHCHECKS, CODEBUILD
	// servicesToBeWhitelist := []string{"S3", "API_GATEWAY"}
	// Comma-serpated env var - no spaces
	servicesToBeWhitelist := strings.Split(os.Getenv("servicesToBeWhitelist"), ",")

	// AWS JSON URL and local download path
	// amazonIPRangesURL := "https://ip-ranges.amazonaws.com/ip-ranges.json"
	amazonIPRangesURL := os.Getenv("amazonIPRangesURL")

	// AWS SSM Param Store that hold the last modified date of the JSON file - format "2020-09-18-21-51-15"
	// previousDateParamStore := "lastModifiedDateIPRanges"
	previousDateParamStore := os.Getenv("previousDateParamStore")

	// AWS DynamoDB table to be created that will maintain a list of all whitelisted IP Ranges
	// dynamoTableName := "whitelistedIPRanges"
	dynamoTableName := os.Getenv("dynamoTableName")

	// Set AWS Region
	// awsRegion := "eu-central-1"
	awsRegion := os.Getenv("awsRegion")

	errMsg := "Error"

	// Parse JSON file into Services data structure
	// var byteValue []byte
	var awsServices Services
	if jsonParseErr := parseJSONfromURL(amazonIPRangesURL, &awsServices); jsonParseErr != nil {
		return errMsg, jsonParseErr
	}

	// Get IP ranges (/16) of all AWS Services that need to be whitelisted
	prefixesForWhitelisting, ipParseErr := parseIPRanges(awsServices, servicesToBeWhitelist)
	if ipParseErr != nil {
		return errMsg, ipParseErr
	}

	// Check how many SGs we need
	sgCheckErr := checkSGCount(securityGroupIDs, prefixesForWhitelisting)
	if sgCheckErr != nil {
		return errMsg, sgCheckErr
	}

	// Check if AWS JSON file was modified since last run
	checkModifiedMsg, jsonModifiedErr := checkIfFileModified(awsServices, previousDateParamStore, awsRegion)
	if jsonModifiedErr != nil {
		return errMsg, jsonModifiedErr
	} else if checkModifiedMsg == "skipRun" {
		fmt.Println("Lambda exeuction completed successfully")
		return fmt.Sprintf("Success"), nil
	}

	// Check if Dynamo table exists
	fmt.Println("Checking if DynamoDB table exists ...")
	tablexists, dbDescribeErr := describeDynamoTable(dynamoTableName, awsRegion)
	if dbDescribeErr != nil {
		return errMsg, dbDescribeErr
	} else if !tablexists {
		fmt.Printf("Table [%v] does not exist\n", dynamoTableName)
		dbCreateErr := createDynamoTable(dynamoTableName, awsRegion)
		if dbCreateErr != nil {
			return errMsg, dbCreateErr
		}
	} else {
		fmt.Printf("[%v] table already exists\n", dynamoTableName)
	}

	var newIPCounter int
	var newIPRanges []string

	fmt.Println("Checking if any IP Ranges need to be whitelisted ...")
	for _, ip := range prefixesForWhitelisting {
		// Check if IP Ranges are already whitelsited
		ipPresent, dbGetErr := getDynamoItem(dynamoTableName, ip, awsRegion)
		if dbGetErr != nil {
			return errMsg, dbGetErr
		}
		if !ipPresent {
			// Update DynamoDB table with IP range that is to be whitelisted
			dbPutErr := putDynamoItem(dynamoTableName, ip, awsRegion)
			if dbPutErr != nil {
				return errMsg, dbPutErr
			}
			newIPRanges = append(newIPRanges, ip)
			newIPCounter++
		}
	}

	if newIPCounter > 0 {
		// Update Security Groups Egress rules
		updateSGErr := updateSecurityGroups(securityGroupIDs, newIPRanges, awsRegion)
		if updateSGErr != nil {
			return errMsg, updateSGErr
		}
		fmt.Printf("Successfully populated IP addresses in security groups %v\n", securityGroupIDs)
	} else if newIPCounter == 0 {
		fmt.Println("No new IP Ranges found, exiting")
		fmt.Println("Lambda exeuction completed successfully")
		return fmt.Sprintf("Success"), nil
	}

	// Describe Security Group Egress rules
	describeSGErr := describeSecurityGroups(securityGroupIDs, awsRegion)
	if describeSGErr != nil {
		return errMsg, describeSGErr
	}
	fmt.Println("Lambda exeuction completed successfully")
	return fmt.Sprintf("Success"), nil
}

// Parse JSON get IP Ranges

func parseJSONfromURL(url string, target interface{}) error {
	var myClient = &http.Client{Timeout: 10 * time.Second}
	r, err := myClient.Get(url)
	if err != nil {
		return fmt.Errorf("parseJSONfromURL: Cannot get JSON from URL : %w", err)
	}
	defer r.Body.Close()

	return json.NewDecoder(r.Body).Decode(target)
}

func parseIPRanges(awsServices Services, serviceWhitelist []string) ([]string, error) {
	var prefixesForWhitelisting []string

	// Go through list of AWS services; get their IP Prefixes if in Whitelist
	for i := 0; i < len(awsServices.Prefixes); i++ {
		// Check if service is to be whitelisted
		if searchStringInArray(awsServices.Prefixes[i].ServiceName, serviceWhitelist) {
			// Parse IP range and make it /16
			ipSlice := strings.Split(awsServices.Prefixes[i].IPPrefix, ".")
			ipFirstTwoOctets := strings.Join(ipSlice[0:2], ".")
			ipRange16 := ipFirstTwoOctets + ".0.0/16"

			// Check if IP subnet range (/16) is already in whitelisted ranges
			if !searchStringInArray(ipRange16, prefixesForWhitelisting) {
				prefixesForWhitelisting = append(prefixesForWhitelisting, ipRange16)
			}
		}
	}

	if len(prefixesForWhitelisting) <= 0 {
		return prefixesForWhitelisting, fmt.Errorf("parseIPRanges: No IP ranges added in list")
	}

	fmt.Printf("IP Ranges that need to be in whitelist: [%v]\n", len(prefixesForWhitelisting))
	fmt.Printf("List of IP Ranges : %v\n", prefixesForWhitelisting)
	return prefixesForWhitelisting, nil
}

// SSM Param Store Functions //

func checkIfFileModified(awsServices Services, previousDateParamStore string, awsRegion string) (responseMsg string, err error) {
	// Check if the AWS JSON file was modified since last run
	previousDate, err := getParamStoreValue(previousDateParamStore, awsRegion)
	if err != nil {
		return "Error", fmt.Errorf("checkIfFileModified: Cannot get SSM Parameter store: %w", err)
	}

	// Type of SSM parameter - String | StringList | SecureString
	ssmParamType := "String"

	// Verify if file has changes since last update
	var currentDate = awsServices.CreationDate
	if previousDate != currentDate {
		fmt.Println("AWS JSON file has changed since last run, previous date: " + previousDate)
		fmt.Println("Updating modified date to " + currentDate)
		// Update Date in SSM Param Store
		setParamStoreValue(previousDateParamStore, currentDate, ssmParamType, awsRegion)
		fmt.Println("Modified date changed")
		return fmt.Sprintf("Success"), nil
	}
	fmt.Println("Last modifed date : " + previousDate)
	fmt.Println("Amazon JSON file has not changed since last run, exiting ...")
	return fmt.Sprintf("skipRun"), nil
}

func getParamStoreValue(previousDateParamStore string, awsRegion string) (string, error) {
	// Create AWS session with default credentials (in ENV vars)
	sess, err := session.NewSession(&aws.Config{
		Region: aws.String(awsRegion)},
	)
	if err != nil {
		return previousDateParamStore, fmt.Errorf("getParamStoreValue: Cannot create AWS config sessions: %w", err)
	}

	// Create an AWS SSM service client
	ssmService := ssm.New(sess, aws.NewConfig())
	// Get SSM Param Store value
	paramKey, err := ssmService.GetParameter(&ssm.GetParameterInput{
		Name:           aws.String(previousDateParamStore),
		WithDecryption: aws.Bool(false),
	})
	if err != nil {
		return previousDateParamStore, fmt.Errorf("getParamStoreValue: Get SSM Parameter: %w", err)
	}
	paramValue := *paramKey.Parameter.Value
	return paramValue, nil
}

func setParamStoreValue(previousDateParamStore string, currentDate string, paramType string, awsRegion string) error {
	// Create AWS session with default credentials(in ENV vars)
	sess, err := session.NewSession(&aws.Config{
		Region: aws.String(awsRegion)},
	)
	if err != nil {
		return fmt.Errorf("setParamStoreValue: Cannot create AWS config sessions: %w", err)
	}

	// Create an AWS SSM service client
	ssmService := ssm.New(sess, aws.NewConfig())
	_, err = ssmService.PutParameter(&ssm.PutParameterInput{
		Name:      aws.String(previousDateParamStore),
		Value:     aws.String(currentDate),
		Overwrite: aws.Bool(true),
		Type:      aws.String(paramType),
	})
	if err != nil {
		return fmt.Errorf("setParamStoreValue: Cannot put SSM Parameter value: %w", err)
	}

	fmt.Printf("Parameter store [%v] type [%v] updated successfully with value [%v]\n", previousDateParamStore, paramType, currentDate)
	return nil
}

// Security Group Functions //

func checkSGCount(securityGroupIDs []string, prefixesForWhitelisting []string) error {
	// Calculate how many SGs are needed to fit all IP Prefixes
	sgAmountNeeded := len(prefixesForWhitelisting)/50 + 1
	sgAmountProvided := len(securityGroupIDs)

	// Check if provided SGs are enough to fit all IP Ranges
	if sgAmountProvided < sgAmountNeeded || sgAmountProvided > sgAmountNeeded {
		errMsg := fmt.Sprintf("checkSGCount: You need [%d] Security Groups, you provided [%d]", sgAmountNeeded, sgAmountProvided)
		return errors.New(errMsg)
	}

	return nil
}

func updateSecurityGroups(securityGroupIDs []string, newIPRanges []string, awsRegion string) error {
	// Create AWS session with default credentials and region (in ENV vars)
	sess, err := session.NewSession(&aws.Config{
		Region: aws.String(awsRegion)},
	)
	if err != nil {
		return fmt.Errorf("updateSecurityGroups: Cannot create AWS config sessions: %w", err)
	}

	// Create an AWS EC2 service client
	svc := ec2.New(sess)

	var counter int

	for _, securityGroup := range securityGroupIDs {
		// Go over each IP range and add it in SG
		for indexKey := range newIPRanges {
			fmt.Printf("\nAdding IP range [%v] to Security Group [%v]... \n", newIPRanges[indexKey], securityGroup)
			// Update Security Group with IP Prefix
			if err := awsUpdateSg(svc, newIPRanges[indexKey], securityGroup); err != nil {
				return fmt.Errorf("updateSecurityGroups: Cannot update security group: %w", err)
			}

			// Max 50 IP ranges per SG
			if counter >= 50 {
				fmt.Printf("updateSecurityGroups: Security group %v full, moving to next one\n", securityGroup)
				counter = 0
				break
			} else {
				counter++
			}
		}
	}
	fmt.Println("Security group update finished")
	return nil
}

func awsUpdateSg(ec2ClientSvc *ec2.EC2, ipForWhitelist string, securityGroup string) error {
	// Define Egress rule Input
	input := &ec2.AuthorizeSecurityGroupEgressInput{
		GroupId: aws.String(securityGroup),
		IpPermissions: []*ec2.IpPermission{
			{
				FromPort:   aws.Int64(443),
				ToPort:     aws.Int64(443),
				IpProtocol: aws.String("tcp"),
				IpRanges: []*ec2.IpRange{
					{
						CidrIp: aws.String(ipForWhitelist),
					},
				},
			},
		},
	}

	// Update Security Group Egress rule from Input
	_, err := ec2ClientSvc.AuthorizeSecurityGroupEgress(input)
	if err != nil {
		if aerr, ok := err.(awserr.Error); ok {
			switch aerr.Code() {
			default:
				return fmt.Errorf("awsUpdateSg: %v", aerr)
			}
		} else {
			// Print the error, cast err to awserr.Error to get the Code and Message from an error.
			fmt.Println(err.Error())
		}
	} else {
		fmt.Println("IP Range added successfully")
	}
	return nil
}

func describeSecurityGroups(securityGroupIDs []string, awsRegion string) error {
	// Create AWS session with default credentials and region (in ENV vars)
	sess, err := session.NewSession(&aws.Config{
		Region: aws.String(awsRegion)},
	)
	if err != nil {
		return fmt.Errorf("describeSecurityGroups: Cannot create AWS config sessions: %w", err)
	}

	// Create an AWS EC2 service client
	svc := ec2.New(sess)

	// Describe current state of Security Groups
	describeResult, err := svc.DescribeSecurityGroups(&ec2.DescribeSecurityGroupsInput{
		GroupIds: aws.StringSlice(securityGroupIDs),
	})
	if err != nil {
		if aerr, ok := err.(awserr.Error); ok {
			switch aerr.Code() {
			case "InvalidGroupId.Malformed":
				fallthrough
			case "InvalidGroup.NotFound":
				exitErrorf("describeSecurityGroups: %s.", aerr.Message())
			}
		}
		exitErrorf("describeSecurityGroups: Unable to get descriptions for security groups, %v", err)
	}

	// Display each security group
	for _, group := range describeResult.SecurityGroups {
		fmt.Println("\nDescribe Security Group Egress Rules : ")
		fmt.Println("\n", group)
	}
	return nil
}

// DynamoDB Functions //

func describeDynamoTable(dynamoTableName string, awsRegion string) (tableExists bool, err error) {
	// Create AWS session with default credentials (in ENV vars)
	sess, err := session.NewSession(&aws.Config{
		Region: aws.String(awsRegion)},
	)
	if err != nil {
		return false, fmt.Errorf("describeDynamoTable: Cannot create AWS config sessions: %w", err)
	}

	// Create a AWS DynamoDB service client
	svc := dynamodb.New(sess)

	input := &dynamodb.DescribeTableInput{
		TableName: aws.String(dynamoTableName),
	}

	_, err = svc.DescribeTable(input)
	if err != nil {
		if aerr, ok := err.(awserr.Error); ok {
			switch aerr.Code() {
			case dynamodb.ErrCodeResourceNotFoundException:
				return false, nil
			case dynamodb.ErrCodeInternalServerError:
				return true, fmt.Errorf("describeDynamoTable: %v", err)
			default:
				return true, fmt.Errorf("describeDynamoTable: %v", err)
			}
		} else {
			// Print the error, cast err to awserr.Error to get the Code and
			// Message from an error.
			fmt.Println(err.Error())
		}
	}
	return true, nil
}

func createDynamoTable(dynamoTableName string, awsRegion string) (err error) {
	// Create AWS session with default credentials (in ENV vars)
	sess, err := session.NewSession(&aws.Config{
		Region: aws.String(awsRegion)},
	)
	if err != nil {
		return fmt.Errorf("createDynamoTable: Cannot create AWS config sessions: %w", err)
	}

	// Create a AWS DynamoDB service client
	svc := dynamodb.New(sess)

	// Define Input for creating table
	input := &dynamodb.CreateTableInput{
		AttributeDefinitions: []*dynamodb.AttributeDefinition{
			{
				AttributeName: aws.String("awsIPRanges"),
				AttributeType: aws.String("S"),
			},
		},
		KeySchema: []*dynamodb.KeySchemaElement{
			{
				AttributeName: aws.String("awsIPRanges"),
				KeyType:       aws.String("HASH"),
			},
		},
		BillingMode: aws.String(dynamodb.BillingModePayPerRequest),
		TableName:   aws.String(dynamoTableName),
	}

	// Create the DynamoDB table using input
	_, err = svc.CreateTable(input)
	if err != nil {
		return fmt.Errorf("createDynamoTable: Cannot create DynamoDB table: %v", err.Error())
	}
	fmt.Printf("createDynamoTable: Successfully create table [%v]\n", dynamoTableName)
	return nil
}

func getDynamoItem(dynamoTableName string, ipRange string, awsRegion string) (ipPresent bool, err error) {
	// Create AWS session with default credentials (in ENV vars)
	sess, err := session.NewSession(&aws.Config{
		Region: aws.String(awsRegion)},
	)
	if err != nil {
		return false, fmt.Errorf("getDynamoItem: Cannot create AWS config sessions: %w", err)
	}

	// Create a AWS DynamoDB service client
	svc := dynamodb.New(sess)

	// Input for Dynamo Get Item operation
	// https://docs.aws.amazon.com/sdk-for-go/api/service/dynamodb/#DynamoDB.GetItem
	input := &dynamodb.GetItemInput{
		Key: map[string]*dynamodb.AttributeValue{
			"awsIPRanges": {
				S: aws.String(ipRange),
			},
		},
		TableName: aws.String(dynamoTableName),
	}

	fmt.Println("Checking item: " + ipRange)

	result, err := svc.GetItem(input)
	if err != nil {
		if aerr, ok := err.(awserr.Error); ok {
			switch aerr.Code() {
			case dynamodb.ErrCodeProvisionedThroughputExceededException:
				return false, fmt.Errorf("getDynamoItem: %v", err)
			case dynamodb.ErrCodeResourceNotFoundException:
				// nil because we dont want it stop executing when item is missing in DB, as we add it later
				return false, nil
			case dynamodb.ErrCodeRequestLimitExceeded:
				return false, fmt.Errorf("getDynamoItem: %v", err)
			case dynamodb.ErrCodeInternalServerError:
				return false, fmt.Errorf("getDynamoItem: %v", err)
			default:
				return false, fmt.Errorf("getDynamoItem: %v", err)
			}
		} else {
			// Print the error, cast err to awserr.Error to get the Code and
			// Message from an error.
			fmt.Println(err.Error())
		}
	}
	stringResult := result.GoString()

	// Regex to match IP address
	re := regexp.MustCompile(`(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)(\.(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)){3}`)
	// Return true if IP Range already in Dynamo table
	return re.MatchString(stringResult), nil
}

func putDynamoItem(dynamoTableName string, ipRange string, awsRegion string) error {
	// Create AWS session with default credentials (in ENV vars)
	sess, err := session.NewSession(&aws.Config{
		Region: aws.String(awsRegion)},
	)
	if err != nil {
		return fmt.Errorf("putDynamoItem: Cannot create AWS config sessions: %w", err)
	}

	// Create a AWS DynamoDB service client
	svc := dynamodb.New(sess)

	// Input for Dynamo Put Item operation
	// https://docs.aws.amazon.com/sdk-for-go/api/service/dynamodb/#DynamoDB.PutItem
	input := &dynamodb.PutItemInput{
		Item: map[string]*dynamodb.AttributeValue{
			"awsIPRanges": {
				S: aws.String(ipRange),
			},
		},
		ReturnConsumedCapacity: aws.String("TOTAL"),
		TableName:              aws.String(dynamoTableName),
	}
	if _, err := svc.PutItem(input); err != nil {
		if aerr, ok := err.(awserr.Error); ok {
			switch aerr.Code() {
			case dynamodb.ErrCodeConditionalCheckFailedException:
				fmt.Println(dynamodb.ErrCodeConditionalCheckFailedException, aerr.Error())
			case dynamodb.ErrCodeProvisionedThroughputExceededException:
				fmt.Println(dynamodb.ErrCodeProvisionedThroughputExceededException, aerr.Error())
			case dynamodb.ErrCodeResourceNotFoundException:
				fmt.Println(dynamodb.ErrCodeResourceNotFoundException, aerr.Error())
			case dynamodb.ErrCodeItemCollectionSizeLimitExceededException:
				fmt.Println(dynamodb.ErrCodeItemCollectionSizeLimitExceededException, aerr.Error())
			case dynamodb.ErrCodeTransactionConflictException:
				fmt.Println(dynamodb.ErrCodeTransactionConflictException, aerr.Error())
			case dynamodb.ErrCodeRequestLimitExceeded:
				fmt.Println(dynamodb.ErrCodeRequestLimitExceeded, aerr.Error())
			case dynamodb.ErrCodeInternalServerError:
				fmt.Println(dynamodb.ErrCodeInternalServerError, aerr.Error())
			default:
				fmt.Println(aerr.Error())
			}
		} else {
			// Print the error, cast err to awserr.Error to get the Code and
			// Message from an error.
			fmt.Println(err.Error())
		}
	}
	fmt.Printf("Adding IP in Dynamo table : %v\n", ipRange)
	fmt.Printf("Successfully updated dynamo table %v\n", dynamoTableName)
	return nil
}

// Helpers //

func exitErrorf(msg string, args ...interface{}) {
	fmt.Fprintf(os.Stderr, msg+"\n", args...)
	os.Exit(1)
}

func searchStringInArray(searchString string, list []string) bool {
	for _, value := range list {
		if value == searchString {
			return true
		}
	}
	return false
}
